### PR TITLE
Bugfix/shadow root

### PR DESCRIPTION
**Pull request by mbgspcii**

### CODE REVIEW

File: documentation/angular.md [hunk 1]
```diff
@@ -199,6 +199,7 @@ installTwicPics( {
 | `anticipation` | [TwicPics](https://www.twicpics.com/) will lazy-load images by default. To avoid too abrupt a transition with elements appearing into view and then images very obviously loading afterwards, [TwicPics](https://www.twicpics.com/) will "anticipate" lazy loading by a factor of the actual viewport. This behavior is controlled by this setting. | `Number` | `0.2` |
 | `domain` | This is your very own [TwicPics domain](https://www.twicpics.com/documentation/subdomain/). Providing it is __mandatory__. | `String` | |
 | `env` | Can be `debug`, `offline` or `production`. When set to `debug`, a gray lightweight `svg` [placeholder](https://www.twicpics.com/docs/api/placeholders) that displays its intrinsic dimensions is displayed in place of all medias targeted by their `src` value. When set to `offline`, these medias are replaced by a simple placeholder that allows to visualise their display area. | `String` | `"production"` |
+| `handlesShadowDom` | Must be set to `true` when using TwicComponents within a shadow DOM. | `boolean` | `false` |
```
Comment (jaubourg) [inline documentation/angular.md:202]: `handleShadowDom` to avoid a unnecessary double `s` (or even `shadowDom` but it may be too short and less obvious).

File: src/_/dom.ts [hunk 2]
```diff
@@ -1,19 +1,21 @@
 export type CreationData = [
```
Comment (jaubourg) [inline src/_/dom.ts:1]: Using an array makes for very convoluted logic.

Wouldn't an object be better there and avoid using the same value for both attributes and shadowDom?

File: src/_/dom.ts [hunk 3]
```diff
@@ -1,19 +1,21 @@
...
         return undefined;
     }
     const [ elementOrTagName, attributes, children ] = data;
     const element =
-        elementOrTagName instanceof HTMLElement ?
-            elementOrTagName :
-            document.createElement( elementOrTagName || `div` );
-    if ( attributes ) {
+        // eslint-disable-next-line no-nested-ternary
+        elementOrTagName instanceof ShadowRoot ? elementOrTagName :
```
Comment (jaubourg) [inline src/_/dom.ts:14]: This is very hard to follow. Why is the expression returning `elementOrTagName` here? It's used as a source for attributes? I really don't understand what's going on and I suspect the handling of the shadowRoot case should be done outside of this DOM-related function.

File: src/_/install.ts [hunk 4]
```diff
@@ -40,6 +42,45 @@ export const configBasedStyle = (): string =>
...
+        const isHost = !parentNode && ( item instanceof ShadowRoot );
+        if ( isHost ) {
+            if ( ( item as unknown as ShadowRoot ).mode === `closed` ) {
+                throw new Error( `cannot use TwicPics components in closed ShadowRoot` );
+            }
+            if ( !styleAdded ) {
+                createElement( [
+                    item as unknown as ShadowRoot,
+                    0,
+                    [ [ `style`, 0, `*STYLE*${ configBasedStyle() }` ] ],
```
Comment (jaubourg) [inline src/_/install.ts:65]: Shouldn't it be `/*STYLE*/`?

File: src/_/install.ts [hunk 5]
```diff
@@ -40,6 +42,45 @@ export const configBasedStyle = (): string =>
 
 export const getDataAttributeName = ( baseName: string ): string => `data-${ config.class }-${ baseName }`;
 
+const markedElements = new WeakSet();
+export const markElementsChain = ( item: Element ): void => {
+    const attributeName = getDataAttributeName( `component` );
```
Comment (jaubourg) [inline src/_/install.ts:47]: This is done at each call to `markElementsChain`. Wouldn't it be better to turn this into a factory that's part of the config itself?

```js
const markElementsFactory = () => {
    const marked = new WeakSet();
    const attributeName = getDataAttributeName( `component` );
    return ( item: Element ): void => {
        /* ... */
    };
};
``` 

That way you only create the set when `handleShadowRoot` is set to `true` at install.

File: src/_/install.ts [hunk 6]
```diff
@@ -40,6 +42,45 @@ export const configBasedStyle = (): string =>
...
+                    0,
+                    [ [ `style`, 0, `*STYLE*${ configBasedStyle() }` ] ],
+                ] );
+                styleAdded = true;
+            }
+            // eslint-disable-next-line no-param-reassign
+            item = ( item as unknown as ShadowRoot ).host;
+        } else {
+        // eslint-disable-next-line no-param-reassign
+            item = ( parentNode as Element );
```
Comment (jaubourg) [inline src/_/install.ts:73]: Unrelated to this whole change but we should probably switch from `parentNode` to `parentElement`.

File: documentation/angular.md (hunk 1 above)
Comment (mbgspcii) [reply documentation/angular.md:202]: - [ ] handleShadowDom -> handleShadowDom

File: src/_/install.ts (hunk 5 above)
Comment (mbgspcii) [reply src/_/install.ts:47]: - [ ] Turn markElementsChain into a factory

File: src/_/install.ts (hunk 4 above)
Comment (mbgspcii) [reply src/_/install.ts:65]: Why ? (search pattern is` /\*STYLE\*/g`)

Comment (jaubourg) [reply src/_/install.ts]: Because I don't follow anything ? :P

File: src/_/dom.ts (hunk 2 above)
Comment (mbgspcii) [reply src/_/dom.ts:1]: - [ ] change CreationData from array to object

File: src/_/dom.ts (hunk 3 above)
Comment (mbgspcii) [reply src/_/dom.ts:14]: - [ ] simplify createElement
